### PR TITLE
feat: auto-generated filetypes metadata

### LINKED ISSUE #383: Better sources discoverability
### Issues

- [X] I have checked [existing issues](https://github.com/jose-elias-alvarez/null-ls.nvim/issues) and there are no existing ones with the same request.

### Feature description

It was possible to get a list of available sources even when they weren't registered, but somewhere between #332 and #321 that ability got removed. It may have never been intended to work that way from the beginning, but it was useful nevertheless. 

Before: 
image

That list is now always empty for non-registered sources, which isn't really that helpful.

### Help

Yes

### Implementation help

There are two options available:
- expose the config so that one can require it without setting up the source, see https://github.com/neovim/nvim-lspconfig/pull/1479
- create a static table that contains metadata about the available sources, see https://github.com/williamboman/nvim-lsp-installer/pull/159

### PR DESCRIPTION
**Pull request by kylo252**
Auto-generate metadata tables for filetype mappings.

Fixes #383

### CODE REVIEW

**Comment by jose-elias-alvarez** [comment]
I'll have a chance to try this out next week, but as far as I can tell, the `diagnostics` and `formatters` tables are not generated automatically, right? That would be a downside, since it means we have to keep each source, its documentation, and the relevant table in sync. Ideally we could generate those automatically by scanning the filesystem.

**Comment by kylo252** [comment]
> I'll have a chance to try this out next week, but as far as I can tell, the `diagnostics` and `formatters` tables are not generated automatically, right? That would be a downside, since it means we have to keep each source, its documentation, and the relevant table in sync. Ideally we could generate those automatically by scanning the filesystem.

I did auto-generate them, but it was a bit fiddly since I did it with `ripgrep` and then a vim-macro. any ideas how to grab them otherwise? It's not like I can source the file, either.

**Comment by jose-elias-alvarez** [comment]
If we're going to use a headless Neovim instance anyways, we could modify the script to use [`vim.loop.fs_scandir`](https://github.com/luvit/luv/blob/master/docs.md#uvfs_scandirpath-callback) to get each path in the `builtins` directory, then scan each of those directories for built-ins, load them, and write their info to a file. 

I'd be hesitant to add this in unless we can automate it as much as possible (meaning the script gets all the necessary information on its own and we can run it via a Git hook or a CI action). We already have an issue with built-in documentation falling out of sync, so I've been thinking about automatically generating that, too.

**Comment by kylo252** [comment]
> If we're going to use a headless Neovim instance anyways, we could modify the script to use [`vim.loop.fs_scandir`](https://github.com/luvit/luv/blob/master/docs.md?rgh-link-date=2021-12-05T18%3A11%3A19Z#uvfs_scandirpath-callback) to get each path in the `builtins` directory, then scan each of those directories for built-ins, load them, and write their info to a file.

Any ideas for how to source these files? Can I just set them all up in headless without opening any file and have access to their configs, or what did you have in mind?

**Comment by kylo252** [comment]
> This is a rough sketch of what I had in mind. It's obviously brittle and we shouldn't use it as-is, but it should illustrate the point:

I didn't see your response in time, but I think my implementation is along the same lines 😄 

---

I should probably add the other sources as well even if the filetype attribute is empty, since you want them for docs generation. Any preference for wildcards for filetypes?
```lua
filetypes = {} 
-- vs 
filetypes = {"*"}
```

**Comment by jose-elias-alvarez** [comment]
Let's go with `filetypes = {}` since that's the current style.

**Comment by kylo252** [comment]
> Let's go with `filetypes = {}` since that's the current style.

Done! 

The new code might be a bit harder to read, but it's pretty generic and helps us avoid hard-coding the method names, just like you wanted :)

File: lua/null-ls/builtins/_meta/filetype_map.lua [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+  html = {
+    formatting = { "rustywind" }
+  },
+  java = {
+    formatting = { "uncrustify" }
+  },
+  javascript = {
+    diagnostics = { "eslint_d" },
+    formatting = { "rustywind" }
+  },
```
Comment (kylo252) [inline lua/null-ls/builtins/_meta/filetype_map.lua:78]: Something got messed up here. I need to make sure to not overwrite old values.

**Comment by jose-elias-alvarez** [comment]
This looks great! Do you think there's a good way to run the script automatically? Since it depends on the Neovim runtime, I imagine setting up CI would be tricky, but maybe a Git hook?

**Comment by kylo252** [comment]
> This looks great! Do you think there's a good way to run the script automatically? Since it depends on the Neovim runtime, I imagine setting up CI would be tricky, but maybe a Git hook?

I think the new workflow worked, but I'll try to give it a test just to make sure.

**Comment by kylo252** [comment]
Oh man, I think I give up on the workflow action. I'm not sure if it's a problem with the authentication, or because the PR is not on the main branch. I'll disable the auto-commit part for now.

**Comment by jose-elias-alvarez** [comment]
@kylo252 Now that this is merged, could we try to fix the workflow? Otherwise I can just remove it for now.

**Comment by kylo252** [comment]
> Now that this is merged, could we try to fix the workflow? Otherwise I can just remove it for now.

Sure! and worst case scenario, you run it as a git-hook. I can recommend [pre-commit](https://github.com/pre-commit/pre-commit) since it's very easy to use. I could even set it up for you if you'd like.

**Pull request merged**